server/src/modules/cats/cats.service.ts

```diff
@@ -0,0 +1,48 @@
+import { Injectable, Inject, NotFoundException } from '@nestjs/common';
+import { Cat } from 'src/models/cat/cat.model';
+import { Mouse } from 'src/models/mouse/mouse.model';
+import { CreateCatDto } from 'src/types/dto/cat.dto';
+import { Op, Sequelize } from 'sequelize';
+
+@Injectable()
+export class CatsService {
+  constructor(
+    @Inject('CATS_REPOSITORY')
+    private catsRepository: typeof Cat
+  ) { }
+
+  async create(createCatDto: CreateCatDto): Promise<Cat> {
+
+    return this.catsRepository.create(createCatDto);
+  }
+
+  async findAll(filter: string): Promise<Cat[]> {
+    if (filter) {
+      const mouseSubQuery = Sequelize.literal(`(SELECT "m"."catId" FROM "mice" as "m" where "m"."name" ILIKE '%${filter}%')`)
+      const whereCondition = {
+        [Op.or]: [
+          { firstName: { [Op.iLike]: `%${filter}%` } },
+          { lastName: { [Op.iLike]: `%${filter}%` } },
+          { id: { [Op.in]: mouseSubQuery } },
+        ],
+      }
+      return this.catsRepository.findAll({
+        where: whereCondition,
+        include: [{ model: Mouse, as: 'mice' }]
+      });
```

Comment:
This is a good idea but an overkill for this use case
you can also do it the following way
```suggestion
      const iLikeQuery = { [Op.iLike]: `%${filter}%` };
      const whereCondition = {
        [Op.or]: [
          { firstName: iLikeQuery },
          { lastName: iLikeQuery },
          { '$mice.name$': iLikeQuery },
        ],
      }
      return this.catsRepository.findAll({
        where: whereCondition,
        include: [{ model: Mouse, as: 'mice' }]
      });
```

Reply:
i did like this before , but gross ask me to try do it with subquery